File: backend/src/aqua/domain/value_objects.py
from dataclasses import dataclass
from enum import Enum, auto

from aqua.domain import errors


@dataclass(kw_only=True, frozen=True)
class Water:
    milliliters: int

    def __post_init__(self) -> None:
        if self.milliliters < 0:
            raise errors.IncorrectWaterAmount()

    def __add__(self, water: "Water") -> "Water":
        return Water(milliliters=self.milliliters + water.milliliters)


@dataclass(kw_only=True, frozen=True)
class WaterBalance:
    class Status(Enum):
        good = auto()
        not_enough_water = auto()
        excess_water = auto()

    water: Water

    def status_when(self, *, target: "WaterBalance") -> Status:
        min_required_quantity = target.water.milliliters - 150
        max_required_quantity = target.water.milliliters + 150

        if self.water.milliliters < min_required_quantity:
            return WaterBalance.Status.not_enough_water

        if max_required_quantity < self.water.milliliters:
            return WaterBalance.Status.excess_water

        return WaterBalance.Status.good


@dataclass(kw_only=True, frozen=True)
class Glass:
    capacity: Water


@dataclass(kw_only=True, frozen=True)
class Weight:
    kilograms: int

    def __post_init__(self) -> None:
        if self.kilograms < 0:
            raise errors.IncorrectWeightAmount()

File: frontend/static/scripts/src/domain/value_objects.ts
import {
    EmptyUsernameError,
    NegativeWaterAmountError,
    FloatWaterAmountError,
    NegativeWeightAmountError,
    FloatWeightAmountError,
} from "./errors.js";

export type UUID = string;

export class Username {
    constructor(readonly text: string) {
        if (text === "")
            throw new EmptyUsernameError();
    }
}

export type JWT = string;

export abstract class PasswordPower {}

export class StrongPasswordPower extends PasswordPower {}

export class WeakPasswordPower extends PasswordPower {
    constructor(readonly reasons: Set<WeakPasswordReasons>) {
        super();
    }
}

export enum WeakPasswordReasons {
    tooShort,
    onlySmallLetters,
    onlyСapitalLetters,
    onlyDigits,
    noDigits, 
}

export class Password<PowerT extends PasswordPower = PasswordPower> {
    private constructor(readonly text: string, readonly power: PowerT) {}

    static with(text: string): Password {
        const weaknessReasons = new Set(Password._weaknessReasonsFor(text));
        let power;

        if (weaknessReasons.size === 0)
            power = new StrongPasswordPower();
        else
            power = new WeakPasswordPower(weaknessReasons);

        return new Password(text, power);
    }

    private static *_weaknessReasonsFor(text: string) {
        if (text.length < 8)
            yield WeakPasswordReasons.tooShort;

        if (text.toLocaleLowerCase() === text)
            yield WeakPasswordReasons.onlySmallLetters;

        if (text.toLocaleUpperCase() === text)
            yield WeakPasswordReasons.onlyСapitalLetters;

        if (Password._hasOnlyDigits(text))
            yield WeakPasswordReasons.onlyDigits;

        if (Password._hasNoDigits(text))
            yield WeakPasswordReasons.noDigits;
    }

    private static _hasOnlyDigits(text: string): boolean {
        return !isNaN(parseInt(text));
    }

    private static _hasNoDigits(text: string): boolean {
        return !text.split('').some(char => char >= '0' && char <= '9');
    }
}

export type StrongPassword = Password<StrongPasswordPower>

export type WeakPassword = Password<WeakPasswordPower>

export class Water {
    constructor(readonly milliliters: number) {
        if (milliliters < 0)
            throw new NegativeWaterAmountError();

        if (!Number.isInteger(milliliters))
            throw new FloatWaterAmountError();
    }
}

export class WaterBalance {
    constructor(readonly water: Water) {}
}

export class Glass {
    constructor(readonly capacity: Water) {}
}

export class Weight {
    constructor(readonly kilograms: number) {
        if (kilograms < 0)
            throw new NegativeWeightAmountError();

        if (!Number.isInteger(kilograms))
            throw new FloatWeightAmountError();
    }
}

export enum WaterBalanceStatus {
    good = 1,
    not_enough_water = 2,
    excess_water = 3,
}
